src/go/k8s/pkg/resources/statefulset_update.go

```diff
@@ -0,0 +1,277 @@
+// Copyright 2021 Vectorized, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+// Package resources contains reconciliation logic for redpanda.vectorized.io CRD
+package resources
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const requeueDuration = time.Second * 10
+
+func (r *StatefulSetResource) updateStsImage(
+	ctx context.Context, sts *appsv1.StatefulSet,
+) error {
+	upgrading := r.pandaCluster.Status.Upgrading
+
+	rpContainer, err := findContainer(sts.Spec.Template.Spec.Containers, redpandaContainerName)
+	if err != nil {
+		return err
+	}
+
+	newImage := fmt.Sprintf("%s:%s", r.pandaCluster.Spec.Image, r.pandaCluster.Spec.Version)
+	if rpContainer.Image != newImage {
+		r.logger.Info("Starting cluster image update", "cluster image", newImage, "container image", rpContainer.Image)
+
+		// Mark cluster as being upgraded.
+		if err := r.updateUpgradeStatus(ctx, true); err != nil {
+			return err
+		}
+	}
+
+	if upgrading {
+		r.logger.Info("Continuing cluster image update", "cluster image", newImage)
+	}
+
+	if upgrading || rpContainer.Image != newImage {
+		podSpec := &sts.Spec.Template.Spec
+		if err := r.modifyPodImage(podSpec, newImage); err != nil {
+			return err
+		}
+
+		if err := r.partitionUpdateImage(ctx, sts, newImage); err != nil {
+			return err
+		}
+
+		// Update is complete for all pods (and all are ready). Set upgrading status to false.
+		if err := r.updateUpgradeStatus(ctx, false); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *StatefulSetResource) updateUpgradeStatus(
+	ctx context.Context, upgrading bool,
+) error {
+	if !reflect.DeepEqual(upgrading, r.pandaCluster.Status.Upgrading) {
+		r.pandaCluster.Status.Upgrading = upgrading
+		if err := r.Status().Update(ctx, r.pandaCluster); err != nil {
+			r.logger.Error(err, "Failed to update RedpandaClusterStatus")
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *StatefulSetResource) partitionUpdateImage(
```

Comment:
I think this logic definitely deserves also unit test coverage

Reply:
As I wrote comment in `updateStsImage` the logic of rolling update should be anchor solely on the following attributes of STS.Status:

- UpdatedReplicas
- CurrentReplicas
- ReadyReplicas
- Replicas

It should find the next ordinal. Reschedule the reconciliation loop. See if it can continue with next ordinal.

Reply:
Rafal, because we are doing partition rolling update, I'm not sure we can use UpdatedReplicas. When we provide partition i, only nodes from i to N-1 will be updated. So the UpdatedReplicas might be misleading. I'd need to check the behavior to be certain. At the same time, since we're contemplating switching to non-statefulset, it'd probably be better to not overly depend on such properties and deal with pods directly during upgrade. This way, if we switch away from StatefulSet the upgrade path will need require fewer changes.

Reply:
Alena, thanks for the suggestion. I will be adding testing, including e2e, once we settle on the implementation. I hope the next few days.

Reply:
I actually reopened this PR to give a similar suggestion as Rafal did, to use the sts status properties :) I think it would be easier and easier to test.

But I get that point with moving out of sts (possibly) ... so I am fine with that. But yeah, actually with this we're replicating a lot of the logic that sts controller is already doing for us. One more reason to not rely on that in the future :-)